pubnub/event_engine/effects.py

```diff
@@ -0,0 +1,431 @@
+            recieve_success = events.ReceiveSuccessEvent(timetoken, region=region, messages=messages)
+            self.event_engine.trigger(recieve_success)
+        self.stop_event.set()
+
+
+class ReconnectEffect(Effect):
+    invocation: invocations.ReconnectInvocation
+    reconnection_policy: PNReconnectionPolicy
+
+    def __init__(self, pubnub_instance, event_engine_instance,
+                 invocation: Union[invocations.PNManageableInvocation, invocations.PNCancelInvocation]) -> None:
+        super().__init__(pubnub_instance, event_engine_instance, invocation)
+        self.reconnection_policy = pubnub_instance.config.reconnect_policy
+        self.max_retry_attempts = pubnub_instance.config.maximum_reconnection_retries
+        self.interval = pubnub_instance.config.RECONNECTION_INTERVAL
+        self.min_backoff = pubnub_instance.config.RECONNECTION_MIN_EXPONENTIAL_BACKOFF
+        self.max_backoff = pubnub_instance.config.RECONNECTION_MAX_EXPONENTIAL_BACKOFF
+
+    def give_up(self, reason: PubNubException, attempt: int, timetoken: int = 0):
+        self.logger.error(f"GiveUp called on Unspecific event. Reason: {reason}, Attempt: {attempt} TT:{timetoken}")
+        raise PubNubException('Unspecified Invocation')
+
+    def failure(self, reason: PubNubException, attempt: int, timetoken: int = 0):
+        self.logger.error(f"Failure called on Unspecific event. Reason: {reason}, Attempt: {attempt} TT:{timetoken}")
+        raise PubNubException('Unspecified Invocation')
+
+    def success(self, timetoken: str, region: Optional[int] = None, **kwargs):
+        self.logger.error(f"Success called on Unspecific event. TT:{timetoken}, Reg: {region}, KWARGS: {kwargs.keys()}")
+        raise PubNubException('Unspecified Invocation')
+
+    def run(self):
+        if self.reconnection_policy is PNReconnectionPolicy.NONE or self.invocation.attempts > self.max_retry_attempts:
+            self.give_up(reason=self.invocation.reason, attempt=self.invocation.attempts)
+        else:
+            attempts = self.invocation.attempts
+            delay = self.calculate_reconnection_delay(attempts)
+            self.logger.warning(f'will reconnect in {delay}s')
+            if hasattr(self.pubnub, 'event_loop'):
+                self.run_async(self.delayed_reconnect_async(delay, attempts))
+
+    async def delayed_reconnect_async(self, delay, attempt):
+        self.stop_event = self.get_new_stop_event()
+        await asyncio.sleep(delay)
+
+        request = Subscribe(self.pubnub).timetoken(self.get_timetoken()).cancellation_event(self.stop_event)
+
+        if self.invocation.channels:
+            request.channels(self.invocation.channels)
+        if self.invocation.groups:
+            request.channel_groups(self.invocation.groups)
+
+        if self.invocation.region:
+            request.region(self.invocation.region)
+
+        if feature_enabled('PN_MAINTAIN_PRESENCE_STATE') and hasattr(self.pubnub, 'state_container'):
```

Comment:
I'm curious, why do you need a check like `hasattr(self.pubnub, 'state_container')`? Isn't internal `state_container` variable always present in `PubNub`?

Reply:
no it's not the internal variable for PubNub. It's only in PubNubAsyncio because it's only used there. This may be redundant check. I'll verify it